### PR TITLE
Patterns: Don't inject theme attr on frontend

### PR DESCRIPTION
**Pull request by ockham**
Result of a [discussion](https://wordpress.slack.com/archives/C02RQBWTW/p1696959836535969?thread_ts=1696880738.627909&cid=C02RQBWTW) with @felixarntz, aiming to fix the performance regression re-introduced by https://core.trac.wordpress.org/changeset/56818.

Needs its [Gutenberg counterpart PR](https://github.com/WordPress/gutenberg/pull/55217), which has the Template Part block fall back to the current theme if no `theme` attribute is present. This should OTOH allow removing `theme` attr injection on the frontend.

To test locally, patch `src/wp-includes/blocks/template-part.php` with the changes from https://github.com/WordPress/gutenberg/pull/55217.diff.

cc/ @felixarntz @gziolo 

Trac ticket: https://core.trac.wordpress.org/ticket/59583

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by felixarntz** [comment]
@ockham In order to be able to benchmark the performance impact, it would be great if you could incorporate the full change into this PR as well, i.e. including the Gutenberg changes from https://github.com/WordPress/gutenberg/pull/55217. Obviously this will cause a CI workflow error and shouldn't be merged, but it makes it easier to compare and discuss the changes.

This is a bit unfortunate, but I think for issues that require work in both Gutenberg and Core, it's the easiest way to holistically reason about the solution. Once we agree on a solution, we can:

1. merge the relevant change in Gutenberg
2. backport the Gutenberg change to Core
3. remove the Gutenberg specific part from here, then refresh with latest `trunk`
4. merge this PR

**Comment by ockham** [comment]
> @ockham In order to be able to benchmark the performance impact, it would be great if you could incorporate the full change into this PR as well, i.e. including the Gutenberg changes from [WordPress/gutenberg#55217](https://github.com/WordPress/gutenberg/pull/55217). Obviously this will cause a CI workflow error and shouldn't be merged, but it makes it easier to compare and discuss the changes.

Makes sense! Done in https://github.com/WordPress/wordpress-develop/pull/5455/commits/62e0be60b443915be6ab9b3635ae36877857b1ca 😊

**Comment by gziolo** [comment]
One open question is whether the Pattern returned from REST API needs to have the current theme injected as the `theme` attribute in the Template Part block included in the pattern. If the answer is yes, then we can probably move the logic that injects it to the REST API controller rather than doing it conditionally in the patterns registry. If the answer is no, we should stop injecting it when accessing all registered patterns.

I asked in the original PR https://github.com/WordPress/gutenberg/pull/53423 that introduced this functionality about some recommendations how we can validate that the functionality still works in the editor after the proposed refactoring.

**Comment by gziolo** [comment]
I'm testing locally with the following diff applied to `trunk`:

```diff
diff --git a/src/wp-includes/blocks/template-part.php b/src/wp-includes/blocks/template-part.php
index a7bd4033af..3ad4009069 100644
--- a/src/wp-includes/blocks/template-part.php
+++ b/src/wp-includes/blocks/template-part.php
@@ -18,13 +18,10 @@ function render_block_core_template_part( $attributes ) {
 	$template_part_id = null;
 	$content          = null;
 	$area             = WP_TEMPLATE_PART_AREA_UNCATEGORIZED;
+	$theme            = isset( $attributes['theme'] ) ? $attributes['theme'] : get_stylesheet();
 
-	if (
-		isset( $attributes['slug'] ) &&
-		isset( $attributes['theme'] ) &&
-		get_stylesheet() === $attributes['theme']
-	) {
-		$template_part_id    = $attributes['theme'] . '//' . $attributes['slug'];
+	if ( isset( $attributes['slug'] ) && get_stylesheet() === $theme ) {
+		$template_part_id    = $theme . '//' . $attributes['slug'];
 		$template_part_query = new WP_Query(
 			array(
 				'post_type'           => 'wp_template_part',
@@ -34,7 +31,7 @@ function render_block_core_template_part( $attributes ) {
 					array(
 						'taxonomy' => 'wp_theme',
 						'field'    => 'name',
-						'terms'    => $attributes['theme'],
+						'terms'    => $theme,
 					),
 				),
 				'posts_per_page'      => 1,
diff --git a/src/wp-includes/class-wp-block-patterns-registry.php b/src/wp-includes/class-wp-block-patterns-registry.php
index e516277e42..ddf73446e7 100644
--- a/src/wp-includes/class-wp-block-patterns-registry.php
+++ b/src/wp-includes/class-wp-block-patterns-registry.php
@@ -165,14 +165,16 @@ final class WP_Block_Patterns_Registry {
 	private function prepare_content( $pattern, $hooked_blocks ) {
 		$content = $pattern['content'];
 
-		$before_block_visitor = '_inject_theme_attribute_in_template_part_block';
+		$before_block_visitor = ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ? '_inject_theme_attribute_in_template_part_block' : null;
 		$after_block_visitor  = null;
 		if ( ! empty( $hooked_blocks ) || has_filter( 'hooked_block_types' ) ) {
 			$before_block_visitor = make_before_block_visitor( $hooked_blocks, $pattern );
 			$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $pattern );
 		}
-		$blocks  = parse_blocks( $content );
-		$content = traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+		if ( ! is_null( $before_block_visitor ) || ! is_null( $after_block_visitor ) ) {
+			$blocks  = parse_blocks( $content );
+			$content = traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+		}
 
 		return $content;
 	}
```

I could confirm in my testing that the `theme` attribute needs to be injected into Template Part blocks in all block patterns when sending to the editor with REST API.

## Benchmarking

> I used the benchmark-server-timing command, with 100 runs each, and I did that 3 times (i.e. 3 medians, each based on 100 runs). All benchmarks without any plugins active (other than Performance Lab for the Server-Timing headers). Below is a summary of the `wp-total` metrics my data.

**With TT4 home page and this patch:**

258.95ms (PR) vs 265.58ms (trunk) --> 2.5% faster 👍🏻 
258.62ms (PR) vs 265.36ms (trunk) --> 2.6% faster 👍🏻 
258.73ms (PR) vs 265.33ms (trunk) --> 2.5% faster 👍🏻

**Comment by gziolo** [comment]
I also tried applying the same mechanism to patterns, as to template, and template parts:

```diff
diff --git a/src/wp-includes/block-template-utils.php b/src/wp-includes/block-template-utils.php
index c5953e1d4c..81f306be93 100644
--- a/src/wp-includes/block-template-utils.php
+++ b/src/wp-includes/block-template-utils.php
@@ -549,15 +549,18 @@ function _build_block_template_result_from_file( $template_file, $template_type
 		$template->area = $template_file['area'];
 	}
 
-	$before_block_visitor = '_inject_theme_attribute_in_template_part_block';
+	$before_block_visitor = ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ? '_inject_theme_attribute_in_template_part_block' : null;
 	$after_block_visitor  = null;
 	$hooked_blocks        = get_hooked_blocks();
 	if ( ! empty( $hooked_blocks ) || has_filter( 'hooked_block_types' ) ) {
 		$before_block_visitor = make_before_block_visitor( $hooked_blocks, $template );
 		$after_block_visitor  = make_after_block_visitor( $hooked_blocks, $template );
 	}
-	$blocks            = parse_blocks( $template_content );
-	$template->content = traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+	if ( ! is_null( $before_block_visitor ) || ! is_null( $after_block_visitor ) ) {
+		$blocks           = parse_blocks( $template_content );
+		$template_content = traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
+	}
+	$template->content = $template_content;
 
 	return $template;
 }
```

When combined with the diff included in https://github.com/WordPress/wordpress-develop/pull/5455#issuecomment-1757470060 I saw better results.

## Benchmarking

> I used the benchmark-server-timing command, with 100 runs each, and I did that 3 times (i.e. 3 medians, each based on 100 runs). All benchmarks without any plugins active (other than Performance Lab for the Server-Timing headers). Below is a summary of the `wp-total` metrics my data.

**With TT4 home page and this patch:**

256.74ms (PR) vs 265.09ms (trunk) --> 3.1% faster 👍🏻 
256.67ms (PR) vs 265.11ms (trunk) --> 3.2% faster 👍🏻 
256.99ms (PR) vs 265.78ms (trunk) --> 3.3% faster 👍🏻

**Comment by felixarntz** [comment]
Given @ockham is out today, I have refreshed this PR with the latest updates from https://github.com/WordPress/gutenberg/pull/55217, as well as the changes suggested above from @gziolo: The `_inject_theme_attribute_in_template_part_block()` should only be called within a REST API request, but not a regular frontend request.

Keep in mind that the changes to the `pattern` and `template-part` blocks are only in here for testing and should be removed prior to commit. They need to be backported from the Gutenberg PR first.

**Comment by felixarntz** [comment]
@ockham I have replaced the tests that are no longer applicable now (https://github.com/WordPress/wordpress-develop/pull/5455/commits/68bfea6c3c68d65cd6c3bf93b2150a26ab02b1a0) with new tests ensuring the template part block still works correctly with the theme attribute (https://github.com/WordPress/wordpress-develop/pull/5455/commits/8522bc87d7389292ade13af166d872cea5f906ad).

Let me know your thoughts, from my perspective this is good to go pending the backport of https://github.com/WordPress/gutenberg/pull/55217.

**Comment by dmsnell** [comment]
Running some tests and I hope they are helpful.

I want to compare three states here:
 - **[control]** before the original refactor in 56805
 - **[broken]** with the refactor in 56805 [git sha 9d039e6c]
 - **[fixed]** with this fix from 56818 [git sha a9bb470f]

Obviously it would be unfair to compare all three of these against each other directly, though, since other improvements may have appeared in `trunk` along the way.

To do this I'm going to create all new commits based on the latest `trunk`.
 - **[control]** is now `trunk` at a8d53284 with reverts of **broken** and **fixed**
   - branch `control` at 76a1aded
 - **[broken]** at `trunk` with revert of **fixed**
   - branch `broken` at 02db6af7
 - **[fixed]** at `trunk`
   - branch `fixed` at a8d53284

- Wiped out the database, reset everything on the site.
- Activate Performance Lab and output buffering plugins, disable all Performance Lab options.
- `curl` the home page 500 times in a row for each branch, switch branches, and repeat until enough samples for all three branches have been collected

<details><summary>performance testing is hard - detour!</summary>



this initial assessment reports that things are insignificantly faster on my test server after the fix in a9bb470f, but interestingly enough there's not that much difference in either of the three branches; plausibly because of the caching that was introduced? I'm guessing that there's a more profound impact when that cache isn't so readily available.

to try and confirm this I reset to the parent commit of 9d039e6c and re-ran the experiment, cherry-picking the fix on that old commit.

 - **[control]** is now `a55dcf46`
 - **[broken]** is now `9d039e6c`
 - **[fixed]** is now `72b829c3`

unfortunately the results are almost identical on these old branches, so I think something was wrong with the tests.

restarting every 500 samples when I change branches, I'm now restoring a fresh copy of the database after activating the plugins, from the control branch.

new graphs based on `trunk` today + the reverts: here are the control, broken, and fixed patches in comparison to each other. these look right.
</details>


block-hooks-patches-today-violins

and here is the same plot if we went back in time to the original parent commit of the breaking patch, then cherry-picked this fix onto the broken patch.

block-hooks-patches-before-violins

so the fixes have less impact today than they had when they were originally applied because `trunk` got faster otherwise. there's only one more interesting feature here I noticed in a plot that deserves exploration: my sneaky suspicion. I'm going to run the tests again but wipe out the database between every request and see if the outliers in these plots (not shown because they peak at around 1400 ms) are outliers or represent the true cost of this stuff when caching isn't available.


<details><summary>But what if the caches aren't primed?</summary>

given that this testing requires resetting the data on every single request, it's very slow and I have set it to switch branches every five requests instead of every five hundred, and I'm going to look at the data before we have nearly as many samples.

block-hooks-no-cache-today-violins

and going back in time…strangely we have two very strong modes that I only noticed after getting about 285 samples of each branch. I don't know why this is. I've separated the graph into a "high" and "low" comparison. maybe this is some `wp-cron` task running? or something random or time-based? whatever it is, the high/slow mode only appears in about 20% of the samples.

block-hooks-no-cache-before-low-violins

block-hooks-no-cache-before-high-violins


</details>

File: src/wp-includes/class-wp-block-patterns-registry.php
Comment (ockham) [inline src/wp-includes/class-wp-block-patterns-registry.php:168]: We might consider moving the code that injects the `theme` attr to the patterns controller instead. Probably okay to leave as-is for now, but might be a nice follow-up post-6.4.

(Noting that if we still insert hooked blocks here, we'd be running `traverse_and_serialize_blocks` twice, which is suboptimal; so we'd also need to move hooked blocks insertion to the controller so we can remove it here; which would then require us to make sure it's run on the frontend. We can probably achieve that through running it in the Pattern block's `render_block`. This means that we'd need to make quite a few changes, which we most likely don't want to do before 6.4.)

Comment (gziolo) [reply src/wp-includes/class-wp-block-patterns-registry.php:168]: Can you explain why `traverse_and_serialize_blocks` would get called twice?

Comment (ockham) [reply src/wp-includes/class-wp-block-patterns-registry.php:168]: > Can you explain why `traverse_and_serialize_blocks` would get called twice?

Assuming that we inject the `theme` attr in the patterns controller, but keep hooked blocks insertion here in the patterns registry (since it's needed both for the frontend and editor (i.e. REST API)) -- both of these call `traverse_and_serialize_blocks` 😅

**Comment by ockham** [comment]
Thanks a lot folks for pushing this along and for the great performance analysis! With https://github.com/WordPress/gutenberg/pull/55217 merged and [ready to be sync'ed](https://github.com/WordPress/wordpress-develop/pull/5468), I'll remove the block code from this PR.

**Comment by ockham** [comment]
> Thanks a lot folks for pushing this along and for the great performance analysis! With [WordPress/gutenberg#55217](https://github.com/WordPress/gutenberg/pull/55217) merged and [ready to be sync'ed](https://github.com/WordPress/wordpress-develop/pull/5468), I'll remove the block code from this PR.

Done in https://github.com/WordPress/wordpress-develop/pull/5455/commits/8394dfd60a675c7dab54e4e86105efd81292b495.

We now need to wait for https://github.com/WordPress/wordpress-develop/pull/5468 to land before we can merge this one 😊

File: src/wp-includes/class-wp-block-patterns-registry.php
Comment (gziolo) [reply src/wp-includes/class-wp-block-patterns-registry.php:168]: I see what you mean here - If we were to run the `theme` attribute injection in the REST API controller. Yes, good point that it would get called twice with different processing 👍🏻

**Review by gziolo — APPROVED** [review]
Everything should be ready to go after https://github.com/WordPress/wordpress-develop/pull/5468 lands and this branch gets rebased with `trunk`.

**Comment by hellofromtonya** [comment]
Heads up: The Gutenberg npm packages are now updated in Core's `trunk` via https://core.trac.wordpress.org/changeset/56849. This changeset includes the package side fixes from https://github.com/WordPress/gutenberg/pull/55217 which were part of https://github.com/WordPress/wordpress-develop/pull/5468.

**Comment by hellofromtonya** [comment]
With the package updates now in `trunk`, I'd suggest rebasing this PR to bring in the latest changes. Then another round of testing just to make sure everything works as expected, i.e. before committing.

**Comment by hellofromtonya** [comment]
Noting (you may already know in which case sorry for my redundant note): The PHPUnit tests are failing with this PR:

```
There was 1 failure:

1) Tests_Blocks_RenderBlockCoreTemplatePart::test_render_block_core_template_part_without_theme_attribute
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'<header class="wp-block-template-part">
-<p>Small Header Template Part</p>
-</header>'
+'Template part has been deleted or is unavailable: small-header'

/var/www/tests/phpunit/tests/blocks/renderBlockCoreTemplatePart.php:60
phpvfscomposer:///var/www/vendor/phpunit/phpunit/phpunit:106
/var/www/vendor/bin/phpunit:118
```

**Comment by ockham** [comment]
Rebased; this should now include the changes to the Pattern and Template Part blocks from https://github.com/WordPress/wordpress-develop/pull/5468, and thus get the PHPUnit tests to pass 🤞

File: src/wp-includes/block-template-utils.php
Comment (felixarntz) [inline src/wp-includes/block-template-utils.php]: Nit-pick: We could use the simpler `null` equality checks here.
```suggestion
	if ( null !== $before_block_visitor || null !== $after_block_visitor ) {
```

File: src/wp-includes/class-wp-block-patterns-registry.php
Comment (felixarntz) [inline src/wp-includes/class-wp-block-patterns-registry.php]: See above:
```suggestion
		if ( null !== $before_block_visitor || null !== $after_block_visitor ) {
```

**Review by felixarntz — APPROVED** [review]
Thanks @ockham for the updates!

Two super minor non-blocking nit-picks, but good to commit IMO!

File: src/wp-includes/block-template-utils.php
Comment (gziolo) [reply src/wp-includes/block-template-utils.php]: Simpler means faster? I coded primarily JavaScript for too long and I’m relearning PHP 😅

Comment (felixarntz) [reply src/wp-includes/block-template-utils.php]: I guess comparing variables is simpler than calling a function, that's all I'm saying.

But since you asked, I got curious: https://3v4l.org/mj5rq 😂 

Seems there is an _extremely_ tiny performance benefit to it too. Really irrelevant, so not a blocker at all. I guess my point is, generally using the variable comparison is encouraged.

Comment (gziolo) [reply src/wp-includes/block-template-utils.php]: Funny enough, in JS I would use `===`, too.

Comment (ockham) [reply src/wp-includes/block-template-utils.php]: I'm fine with this change. Personally, I might lean towards

```suggestion
	if ( $before_block_visitor || $after_block_visitor ) {
```

but I don't mind the `!== null` checks either 🤷‍♂️

**Comment by ockham** [comment]
Committed to Core in https://core.trac.wordpress.org/changeset/56896.

File: src/wp-includes/block-template-utils.php
Comment (dmsnell) [reply src/wp-includes/block-template-utils.php]: it might be worth pointing out caution when relying on these kinds of micro-benchmarks. this is not only because they skew the performance picture, since we're taking the code outside of the context in which it's run, but it's easy to accidentally misrepresent what we want to test.

what we actually tested in the link above is whether it's faster to do one thing by itself, or do that same thing and then do another thing. in this case, `null === $v` and `is_null( $v )` both compile into the same literal operation, a PHP opcode called `TYPE_CHECK` as you can see in the VLD output of the code. the difference being in the two constructions is that `null !== $v` performs the type check and then calls JMPZ directly, whereas `! is_null( $v )` negates the return value before calling JMPZ. why does it negate it? because we asked it to. the first branch asks PHP "tell me if this value is anything but NULL" while the second branch asks it "tell me if this value is NULL." it was able to recognize the optimization in the first case but not the second, because we've added an external data dependency on the return value through the `!`.

if you instead compare `null === $v` with `is_null( $v )` you will see that they are literally identical and there's no performance win either way. in both of these cases it asks PHP directly, "tell me if this value is NULL type."

https://3v4l.org/Daah2

in fact, look at what happens if we change `if ( ! is_null() )` into `if ( is_null() ) {} else { … }`!

https://3v4l.org/QiHIA

the two are the same speed again. but we shouldn't take away from this that…

```php
if ( is_null( $v ) ) {
    // WordPress ricer.
} else {
    do_something();
}
```

…is how we should code. I think we should take away from this that synthetic microbenchmarks are extremely unreliable indicators of code performance.

in fact, one more example of where these things can go wrong directly shows how performance can change when there is more context available. instead of calling `null !== $v` let's call `null !== $v && $v > 5`

https://3v4l.org/NJRum

now the `null !== $v` is only faster than `! is_null( $v )` some of the time, and statistically there's not sufficient evidence to demonstrate that the difference we measure isn't explained by the noisiness of our testing environment. (if you aren't seeing each branch trade places for which is faster, re-load the page and run the experiment multiples times; you should see flips).

once we add more code or real code then things can be funny in the CPU, things can happen out of order, conditions inside expressions can be pre-computed and available by the time we "get" to the `if` statement, and the more actual stuff that's happening, the harder it is to predict what matters in situations like this. micro-synthetic benchmarks are great at showing situations we will never see in production and often only show their results the way they do because we've constructed a situation specifically to lead to that outcome, unintentionally.

---

all I meant through this digression is that I support coding first to what feels most expressive and then only change that if there's is a compelling argument from intuition, experience, or measurements that the code in question behaves in a way that merits optimizations.

there's a note in the WordPress coding standards and in the PHP docs (in a comment) about pre-increment being faster than post-increment and that's just demonstrably not true and it's even more trivial than here with `! is_null()` to find those counter cases.

this specific case is fairly non-controversial, because they both seem fine to me, and I even prefer `null !==` for my own style, but this is what it is - a style argument, not a performance argument. it _could_ be different if the distinction is between calling a function or an opcode, but here the generated code in both situations is nearly identical except for a negation.

thanks everyone for being non-hyped about this. I know you've all shared even-keeled thoughts and nobody is expressing strong claims or controversy; that's the same tone I'm trying to share here. only sharing this long comment because I've seen this run to micro-benchmarks multiple times and I hope that we can all understand that they rarely reflect the performance implications of the code we discuss, that they give the wrong impression just as often as they give the right one, unless the performance impact is so great that there's no wiggle room.

Comment (felixarntz) [reply src/wp-includes/block-template-utils.php]: 100% agreed. I think such feedback is often purely code style preference rather than performance related (at least it is the case here), which is why I clarified this to not be a blocker at all. IMO not calling a function for something as simple as a strict variable comparison makes sense to me as a coding principle, but that is not driven by performance.

There are a few instances where such nit-picks have more of a performance background (e.g. using `isset()` over `array_key_exists()`), but even there the impact is negligible. On that example I would feel slightly stronger about it (since `isset()` is a language construct and thus encouraged over `array_key_exists()`) - but realistically even that is certainly not going to make or break performance in a PR, it's more of a best practice.

All that is to say: I fully agree with you, and most such comments should be addressed based on preferences between the PR author and reviewer, but not be the foundation for blocking a PR from commit.

Comment (dmsnell) [reply src/wp-includes/block-template-utils.php]: thanks @felixarntz - I do want to affirm what you're saying though, and note here that `is_null()` is also not an actual function but another language construct, and that's why these two styles are approximately equal. it makes me sad that in certain cases function calls still have high overhead.

sometimes the minutia does matter though. replacing arrays with a "record" style class in the Tag Processor led to an incredible performance improvement. thankfully in these cases a real impact appears in realistic testing environments.

**edit** the funnier thing to me is not that we're comparing a function call against equality, but in this case it's the `!==` which isn't what it appears to be, as that equality is hiding a call to an internal PHP function to return the type of the variable, matching the behavior of `is_null()` because it's faster than testing for equality 🙃

**Pull request closed without merging**
